### DISCUSSION TITLE
Integration Checklist

### DISCUSSION THREAD

**Discussion by aidnem** [team401/high-key-2024#27] [General]
List of things we should integrate, and in what order (OPEN TO SUGGESTIONS/REVISION):
1. [x] Swerve drive module PID tuning + verification
2. [x] Verify driving works in teleop
3. [x] Tune PID for arm
4. [x] Tune PID for shooter RPM
5. [x] Tune PID for aiming robot at speaker
6. [x] Verify intake works + banner sensor stops intake
7. [x] Verify end to end scoring with fixed angle/RPM target (pick up note, score note). The fixed angle can be up against the subwoofer for max repeatability
8. [x] Initial verification of localization (vision + odometry) based on manual driving around the field
9. [x] Build out lookup table for arm angle and shooter RPM based on distance from speaker
10. [ ] Path following control by driving auto routes (determine speed/accel limits for auto, retune modules if needed)
11. [ ] Implement + tune autonomous
12. [ ] LEDs for beelink not on
13. [ ] LEDs for note indicator

**Comment by aidnem** [comment]
@jkleiber You probably have more of a vision in mind for what integration will look like, feel free to edit this post or suggest edits as you see fit. I just wanted a good way to have an overall view of the integration process and track progress.

**Reply by jkleiber** [reply]
This is a good list, but the order should be based on the following:
1. Technical risk (how difficult do we think a task will be to execute from an engineering standpoint)
2. Time estimate
3. Pre-requisites

So I would propose the following ordered list:
1. Swerve drive module PID tuning + verification
2. Verify driving works in teleop
3. Tune PID for arm
4. Tune PID for shooter RPM
5. Tune PID for aiming robot at speaker
6. Verify intake works + banner sensor stops intake
7. Verify end to end scoring with fixed angle/RPM target (pick up note, score note). The fixed angle can be up against the subwoofer for max repeatability
8. Initial verification of localization (vision + odometry) based on manual driving around the field
9. Build out lookup table for arm angle and shooter RPM based on distance from speaker
10. Path following control by driving auto routes (determine speed/accel limits for auto, retune modules if needed)
11. Implement + tune autonomous
12. LEDs for beelink not on
13. LEDs for note indicator

**Comment by PChild** [comment]
LEDs at some point. 🥺

https://github.com/user-attachments/assets/8def165e-5efe-4525-ba46-5a2c3303fc5b

**Comment by jkleiber** [comment]
Update: intake works but we need to make the banner sensor work and only allow full intake when aimer is down

**Comment by jkleiber** [comment]
@aidnem @linglejack06 @avidraccoon @minhnguyenbhs 

Would Saturday shop from 10am to 3pm work well for y'all? If not, what time frame would be better?

I don't expect that everyone will be able to make 100% of this time, but I just want to pick a time where we can have continuous attendance and get a lot done

**Reply by avidraccoon** [reply]
I'm sorry but I won't be able to go to the shop on Saturday.

**Reply by minhlpnguyen** [reply]
hi! I'm so sorry but something came up and I don't know if I'll be able to come tomorrow

**Reply by minhlpnguyen** [reply]
@linglejack06 would it maybe be possible for me to ask you for a ride? I'm not sure if I can go either way but would that be too much trouble for you? thanks- sorry

**Reply by aidnem** [reply]
I can make it sometime around 10 but unfortunately can't give anybody a ride, sorry.

**Comment by jkleiber** [comment]
I am writing this message to try to provide some direction on what we should be aiming to accomplish over the next two shops for all our Robots. 

Robot assignments - please make sure the robot you are assigned to is ready for competition:
- High Key: @aidnem @linglejack06 
- Apep: @minhnguyenbhs 
- C bot: @avidraccoon 
If your work is done, then I'll leave it to your discretion as to which robot needs the most support.

**Monday October 14**
High Key: fully functioning teleop by end of shop
- [x] Shooter lookup table
- [x] Drivetrain auto aim at speaker

Getting the drivetrain auto aim should be easy if vision is working well. The main priority of today is getting a highly accurate lookup table for shooter so we can do auto on Wednesday.

Apep:
- [ ] Test auto a few times to make sure it's mostly working
- [x] Do some teleop if time allows

C bot priorities
- [ ] If the superstructure is on, test/integrate scoring. The robot is drivable (I fixed the steering offsets on Saturday)
- [ ] See if you can get driving working without the weird control issues once scoring is done or if the superstructure isn't on (steers halfway and then only gets to the setpoint if you stop commanding things)

**Wednesday October 16**
High Key:
Build out autonomous incrementally in the following order:
- [ ] Score the preload and sit still (1 note)
- [ ] Center start 4 note wing auto (preload, wing notes)
- [ ] Center start 3 note (preload, center wing note, 2 centerline notes)
- [ ] Center start 4 note (see above but go get another note from center)
- [ ] Amp side start 3 note (preload, amp side wing note, centerline note)
- [ ] Amp side 4 note (see above but go get another note from center)
- [ ] Source side 3 note (preload, 2 from centerline far side)

We want to have an auto at each start position in the following preference order: center, amp, source. Bare minimum is that we need to score preload. 

If vision is working well, then **prioritize picking up notes rather than scoring them**. If we had more time we would want high shooter accuracy, but theoretically the shooter lookup table is correct from Monday. So if an auto is consistently picking up all the notes and shooting them at the goal move on to the next one to tune the path. If there's time on Friday or at competition we can mess with the lookup table.

Apep and C Bot: 
Finish anything that didn't get done on Monday. Driver practice if High Key isn't in the way.

cc: @PChild for awareness on what programming should be doing today/Wednesday

**Comment by linglejack06** [comment]
@jkleiber @aidnem do we have a good way to quickly adjust shooter rpm? If not I thought of a basic hack we could do to prevent redeploying after every rpm change

**Reply by jkleiber** [reply]
We should develop a test mode to set aimer angle and shooter RPM using elastic

**Comment by linglejack06** [comment]
@jkleiber @aidnem after several failures auto works with alignment now, shooting preload is fairly accurate. It gets to intake the first note but the arm isn't at intake position so it intakes into robot above arm. I'm gonna work on adding a named command to make arm go to intake position tonight. Also , by the time robot got out to the first notes position, vision estimate was completely wrong (likely due to lack of apriltags on our field by amp). Therefore the alignment was really messed up. But I think with proper comp field vision should be sufficient to aim from out there.

**Reply by jkleiber** [reply]
Thanks @linglejack06 , it sounds like we need to do a better job of rejecting bad vision estimates so we can be robust to the case where we don't have much tag information. I think that should at least help us get all the wing notes since the speaker tags will be well in view

**Comment by jkleiber** [comment]
@aidnem / @linglejack06 can we work to get the 14-calibrate-cameras branch merged to main today? Then we should create a competition branch + release on Github prior to BRB

**Reply by linglejack06** [reply]
@jkleiber yes, do you want me to add the arm intake position to this branch before we merge?

**Reply by jkleiber** [reply]
Let's use a new branch

**Reply by linglejack06** [reply]
Ok then 14-calibrate cameras should be able to be reviewed and merged

**Reply by jkleiber** [reply]
It still needs a pull request

**Comment by linglejack06** [comment]
I’ll open that when I’m by my Mac Sent from my iPhoneOn Oct 17, 2024, at 1:42 PM, Justin Kleiber ***@***.***> wrote:﻿
It still needs a pull request

**Comment by minhlpnguyen** [comment]
hi I’m feeling a little unwell today, I’m not sure if I’ll be able to go to BRB but I’ll do my best to get there asap 
if I can't go, I hope things go well today!!

**Reply by aidnem** [reply]
@minhnguyenbhs Sorry to hear that, I hope you feel better! You didn't miss much lol

**Comment by jkleiber** [comment]
@aidnem I think we should baseline the following overall goal for B bot this week: "make the robot great in teleop"

This doesn't mean we are going to abandon auto, nor does it mean that we should make the robot less automated. It just means that we should prioritize things like ease of operation.

The following will accomplish that goal:
1. Making auto alignment work
2. Tuning all controllers so that "force shoot" is not required
3. Making amp warmup / scoring process seamless
4. Adding a "pass shot" to feed notes to our alliance wing
5. Adding LED indication for note possession / beelink not on

Once these things are complete, we can resume our incremental build up towards autonomous.

**Reply by aidnem** [reply]
I agree. I believe that auto alignment was the culprit behind force shoot being required, which would mean that (hopefully) if we fix align it will fix both issues. Fixing the aimer is also going to be crucial so that we can actually make shots. Also, I think localization was a bit of an issue (we can't aim if we don't know where we are).
Would you mind elaborating on what you mean by making amp process seamless?

**Reply by jkleiber** [reply]
Currently we have an untested amp mode (besides shooting notes in the pits to verify logic). Last year scoring in the amp was possible in theory but we were so inefficient with it that it wasn't worth it. I want to ensure that it's as easy to score in the amp as it is to score in the speaker (and for both of these things to be easy overall)

For localization I think we should start by only considering 1 camera and then incrementally adding cameras as we feel confident in the ones that have been incorporated

**Comment by linglejack06** [comment]
What went wrong with the alignment? @jkleiber @aidnem it was working great at shop

**Reply by jkleiber** [reply]
It just straight up did not work at BRB as far as I know. This could have been a vision problem. In auto we would either:
1. Not move at all
2. spin in circles

**Reply by aidnem** [reply]
@jkleiber I think it was aligning properly in auto, I suspect that when I flipped all of the aligns by 180 degrees to make it aim at the speaker (instead of shooting the note at the referee) I forgot to update the [isDriveAligned function](https://github.com/team401/high-key-2024/blob/9f54fdc250bc8416514e77f87ed92fb9805fd707/src/main/java/frc/robot/subsystems/drive/PhoenixDrive.java#L323). This means that it would align correctly but then think that it wasn't aligned, causing us to never auto-shoot, and requiring all mashers to force shoot.

**Reply by jkleiber** [reply]
@aidnem can you create a ticket and a branch to fix that? If you have time it would be good to try it in the simulator to see if that works

**Reply by aidnem** [reply]
Issue created: #56.

**Reply by aidnem** [reply]
Also @jkleiber we were never moving in auto because it would aim and spin up flywheels but never shoot, and the auto was waiting for the shot to be taken before driving away. Supposedly this will go away once we satisfy it that it can take shots without force shoot.

**Comment by linglejack06** [comment]
I noticed that would happen in auto when vision estimate pose was wildly different than real pose. As long as no changes were made to the alignment part that’s the only thing I think that could be wrong. I guess we will look at it at shop tomorrow and try to figure it outSent from my iPhoneOn Oct 20, 2024, at 3:54 PM, Justin Kleiber ***@***.***> wrote:﻿
It just straight up did not work at BRB as far as I know. This could have been a vision problem. In auto we would either:

Not move at all
spin in circles

**Reply by aidnem** [reply]
Probably a vision issue then. I think we should start with the most reliable camera and then see if we can get just that camera to work at all.

**Comment by linglejack06** [comment]
yeah, i had adjusted the speaker alignment specifically by pi radians (as
in shop it was pointed the wrong way), this was then accounted for in the
is drive aligned method by adding back the pi radians. but if you
changed the getAlignment function for speaker then the is drive aligned
would be messed up

**Reply by aidnem** [reply]
Yes, for some reason the robot turned 180 degrees away from the speaker and shot into the crowd during the first match's auto, so I inverted the rotation after that.

**Comment by linglejack06** [comment]
Was that on blue alliance? If so it’s because I forgot to add a conditional. On red alliance 180 degrees has to be added to the heading to align to speaker, but blue alliance works without the heading. So the conditional should say if red alliance add 180 and if red alliance subtract the 180 for the is drive aligned method. If blue alliance then the rotation is left alone for the is drive aligned and the get rotation Sent from my iPhoneOn Oct 21, 2024, at 12:52 PM, aidnem ***@***.***> wrote:﻿
Yes, for some reason the robot turned 180 degrees away from the speaker and shot into the crowd during the first match's auto, so I inverted the rotation after that.

**Reply by aidnem** [reply]
Yes that was on blue, that makes sense. Any idea why the headings are wrong by 180 degrees on red?

**Comment by linglejack06** [comment]
I assume it’s due to the coordinate plane of the field somehow. I can explain better with a drawing at shopSent from my iPhoneOn Oct 21, 2024, at 2:13 PM, aidnem ***@***.***> wrote:﻿
Yes that was on blue, that makes sense. Any idea why the headings are wrong by 180 degrees on red?

**Comment by linglejack06** [comment]
Yes the wait command specifically waits until the note has left the shooter before doing anything. So the auto will not continue unless the note is shot or spit outSent from my iPhoneOn Oct 21, 2024, at 2:17 PM, aidnem ***@***.***> wrote:﻿
Also @jkleiber we were never moving in auto because it would aim and spin up flywheels but never shoot, and the auto was waiting for the shot to be taken before driving away. Supposedly this will go away once we satisfy it that it can take shots without force shoot.

**Comment by linglejack06** [comment]
Also it won’t continue on path if isDriveAligned is false so two issues thereSent from my iPhoneOn Oct 21, 2024, at 2:17 PM, aidnem ***@***.***> wrote:﻿
Also @jkleiber we were never moving in auto because it would aim and spin up flywheels but never shoot, and the auto was waiting for the shot to be taken before driving away. Supposedly this will go away once we satisfy it that it can take shots without force shoot.

**Comment by jkleiber** [comment]
**Plan for Shop Today**

- [x] Test / tune pass shot
- [x] Test seamless amp
- [x] Get better at shooting in teleop (possibly vision tuning)
- [x] If time allows, run some simple auto

**Reply by aidnem** [reply]
This is probably obvious but in case I forget later, tuning-shot mode will work for pass testing so that we don't need it to have its own test mode.

**Comment by aidnem** [comment]
@jkleiber want me to create an issue for closed-loop path following so that if we have time for autos we can make them actually good?

**Reply by jkleiber** [reply]
Let's use #35

**Comment by aidnem** [comment]
@PChild When arm is being controlled to the bottom position, there's enough play in the chain for the aimer to swing up due to acceleration of the robot. There's just enough play in the system for this to sometimes intake notes into the middle of the robot if we intake while accelerating just right. Is there any way to reduce backlash in the chain? My proof that this is not a code issue is that, when setting the motor to brake mode, I can lift the arm a little bit before I feel the brake mode resistance as the motor actually starts to turn.

**Reply by PChild** [reply]
We can try tensioning the chain more, I don't think we have time to shim gears in the gearbox now. Not sure if it's realistic to implement at this time but the other thing we could do is drive the arm into the hard stop with a low current limit, this should take all of the backlash out of the chain and gear train

**Reply by PChild** [reply]
Also we should check that when the robot is enabled, behavior of brake mode when disabled != PID control behavior, the encoder is 1:1 on the arm so it shouldn't see any lash from chain or gears.

**Reply by aidnem** [reply]
This is true, PID might just be controlling down too weak to bring the arm back down immediately. We don't really have a good way to gently drive into the hardstop but I feel like it wouldn't be enough anyway unless we were driving hard enouhg for it to heat the motor up a lot and probably be bad.
We could consider a fancy feedforward solution based on acceleration but honestly we could also just tell oliver not to accelerate too hard while intaking.

**Reply by PChild** [reply]
Adjusting the current limit lets you control both how much torque the system pushes into the hard stop with and how much heat is generated. The driver should be allowed to accelerate as hard as they can.

**Reply by jkleiber** [reply]
@aidnem we could use @minhnguyenbhs work in #47 to apply a constant downward aimer output and a low current limit in intake mode

**Reply by aidnem** [reply]
Good idea, I was thinking of something similar.
Do we know if a current limit will do what we want (not disagreeing, just curious)? I'm not sure how smart it is, but does it just zero out motor voltage when current goes above the limit and then wait for a period of time?
